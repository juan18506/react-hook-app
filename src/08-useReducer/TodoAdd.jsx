import PropTypes from 'prop-types';
import { useForm } from '../hooks/useForm';

export const TodoAdd = ({ onNewTodo }) => {
  const { formState, handleInputChange, handleResetForm } = useForm({
    description: '',
  });
  
  const { description } = formState;
  
  const handleFormSubmit = ( event ) => {
    event.preventDefault();
    if ( description.length <= 1 ) return;

    const newTodo = {
      id: new Date().getTime(),
      description: description,
      done: false,
    }

    onNewTodo( newTodo );
    handleResetForm()
  }

  return (
    <form onSubmit={ handleFormSubmit }>
      <input 
        type="text" 
        placeholder="description"
        name="description"
        className="form-control"
        value={ description }
        onChange={ handleInputChange }
      />

      <button 
        type="submit"
        className="btn btn-outline-primary mt-1"
      >
        Add
      </button>
    </form>
  )
}

TodoAdd.propTypes = {
  onNewTodo: PropTypes.func.isRequired,
}