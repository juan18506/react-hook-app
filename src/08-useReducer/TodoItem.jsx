import PropTypes from 'prop-types';

export const TodoItem = ({ todo, onRemoveTodo, onToggleTodo }) => {
  return (
    <li className="list-group-item d-flex justify-content-between">
      <span 
        className={`align-self-center ${ (todo.done) ? 'text-decoration-line-through' : '' }`}
        onClick={ () => onToggleTodo( todo.id ) }
      >
        { todo.description }
      </span>

      <button 
        className="btn btn-danger" 
        onClick={ () => onRemoveTodo( todo.id ) }
      >
        Delete
      </button>
    </li>
  )
}

TodoItem.propTypes = {
  todo: PropTypes.shape({
    id: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
    done: PropTypes.bool.isRequired,
  }).isRequired,
  onRemoveTodo: PropTypes.func.isRequired,
  onToggleTodo: PropTypes.func.isRequired,
}