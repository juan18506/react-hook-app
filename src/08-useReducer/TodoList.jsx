import PropTypes from 'prop-types';
import { TodoItem } from './'

export const TodoList = ({ todos = [], onRemoveTodo, onToggleTodo }) => {
  return (
    <ul className="list-group">
    {
      todos.map(( todo ) => (
        <TodoItem 
          key={ todo.id } 
          todo={ todo } 
          onRemoveTodo={ onRemoveTodo }
          onToggleTodo={ onToggleTodo }
        />
      ))
    }
  </ul>
  )
}

TodoList.propTypes = {
  todos: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      description: PropTypes.string.isRequired,
      done: PropTypes.bool.isRequired,
    })
  ).isRequired,
  onRemoveTodo: PropTypes.func.isRequired,
  onToggleTodo: PropTypes.func.isRequired,
}