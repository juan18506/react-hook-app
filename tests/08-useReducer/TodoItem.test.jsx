import { describe, it, expect, afterEach, vi } from 'vitest';
import { screen, render, cleanup, fireEvent } from '@testing-library/react';
import { TodoItem } from '../../src/08-useReducer/TodoItem';

const onRemoveTodoMock = vi.fn();
const onToggleTodoMock = vi.fn();

describe('Tests in <TodoItem />', () => {
  const todo = {
    id: 1,
    description: 'todo 1',
    done: false,
  };
  
  afterEach(() => {
    vi.clearAllMocks();
    cleanup();
  });  

  it('should match snapshot', () => {
    const { container } = render( 
      <TodoItem 
        todo={ todo } 
        onRemoveTodo={ onRemoveTodoMock } 
        onToggleTodo={ onToggleTodoMock } 
      /> 
    );

    expect( container ).toMatchSnapshot()
  });

  it('should show default values', () => {
    render( 
      <TodoItem 
        todo={ todo } 
        onRemoveTodo={ onRemoveTodoMock } 
        onToggleTodo={ onToggleTodoMock } 
      /> 
    );

    expect( screen.getByText(todo.description) ).toBeTruthy();
    expect( screen.getByText(todo.description).className ).not.toContain( 'text-decoration-line-through' );
    expect( screen.getByRole('button', { name: 'Delete' }) ).toBeTruthy();
  });

  it('should call onToggleTodo', () => {
    render( 
      <TodoItem 
        todo={ todo } 
        onRemoveTodo={ onRemoveTodoMock } 
        onToggleTodo={ onToggleTodoMock } 
      /> 
    );
    fireEvent.click( screen.getByText(todo.description) );

    expect( onToggleTodoMock ).toHaveBeenCalledWith( todo.id );
  });

  it('should call onRemoveTodo', () => {
    render( 
      <TodoItem 
        todo={ todo } 
        onRemoveTodo={ onRemoveTodoMock } 
        onToggleTodo={ onToggleTodoMock } 
      /> 
    );
    fireEvent.click( screen.getByRole('button', { name: 'Delete' }) );

    expect( onRemoveTodoMock ).toHaveBeenCalledWith( todo.id );
  });

  it('should add a class if todo is done', () => {
    const todo = {
      id: 1,
      description: 'todo 1',
      done: true,
    };

    render( 
      <TodoItem 
        todo={ todo } 
        onRemoveTodo={ onRemoveTodoMock } 
        onToggleTodo={ onToggleTodoMock } 
      /> 
    );

    expect( screen.getByText(todo.description).className ).toContain( 'text-decoration-line-through' );
  });
});
