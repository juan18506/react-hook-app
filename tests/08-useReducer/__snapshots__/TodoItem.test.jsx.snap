// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests in <TodoItem /> > should match snapshot 1`] = `
<div>
  <li
    class="list-group-item d-flex justify-content-between"
  >
    <span
      class="align-self-center "
    >
      todo 1
    </span>
    <button
      class="btn btn-danger"
    >
      Delete
    </button>
  </li>
</div>
`;
